src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -687,3 +688,134 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
+        i += 1
+      }
+      // if last parent partition overlaps next blockRow, don't advance
+      if (parentPartitionBoundaries(i) > lastRowInBlock)
+        i -= 1
+      
+      deps(blockRowIndex) = ab.result()
+      blockRowIndex += 1
+    }
+
+    deps
+  }
+}
+
+class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, blockSize: Int) extends RDD[Int](irm.rows.sparkContext, Nil) {
+  private val rows = irm.numRows()
+  private val cols = irm.numCols()
+  private val parentPartitions = irm.rows.partitions
+  private val parentPartitionBoundaries: Array[Long] = irm.rows.countPerPartition().scanLeft(0L)(_ + _)
+  
+  // all IndexedRows must be present in RDD
+  assert(rows == parentPartitionBoundaries.last,
+    s"IndexedRowMatrix has $rows rows but RDD only has ${parentPartitionBoundaries.last} IndexedRows.")
+
+  private val gp = GridPartitioner(blockSize, rows, cols)
+  private val blockRowDependencies = WriteBlocksRDD.computeBlockRowDependencies(parentPartitionBoundaries, gp)
+  private val truncatedBlockRow = rows / blockSize
+  private val truncatedBlockCol = cols / blockSize
+  private val excessRows = (rows % blockSize).toInt
+  private val excessCols = (cols % blockSize).toInt
+  
+  private val d = digitsNeeded(gp.numPartitions)
+  private val sHadoopBc = irm.rows.sparkContext.broadcast(
+    new SerializableHadoopConfiguration(irm.rows.sparkContext.hadoopConfiguration))
+
+  override def getDependencies: Seq[Dependency[_]] = {  
+    Array[Dependency[_]](
+      new NarrowDependency(irm.rows) {
+        def getParents(partitionId: Int): Seq[Int] = blockRowDependencies(partitionId)
+      }
+    )
+  }
+  
+  protected def getPartitions: Array[Partition] =
+    (0 until gp.rowPartitions).map(IntPartition).toArray[Partition]
+  
+  def compute(split: Partition, context: TaskContext): Iterator[Int] = {
+    val blockRowIndex = split.index
+    val firstRowInBlock = blockRowIndex * blockSize
+    val nRowsInBlock = if (blockRowIndex == truncatedBlockRow) excessRows else blockSize
+    val lastRowInBlock = firstRowInBlock + nRowsInBlock // technically, first row in next blockRow
+    
+    val data0 = Array.ofDim[Double](nRowsInBlock * blockSize)
+    
+    var blockColIndex = 0
+    while (blockColIndex < gp.colPartitions) {
+      val firstColInBlock = blockColIndex * blockSize
+      
+      val nColsInBlock = if (blockColIndex == truncatedBlockCol) excessCols else blockSize
+      var offset = 0
+      blockRowDependencies(blockRowIndex).foreach { parent =>
+        val firstRowInParent =  parentPartitionBoundaries(parent)
+        val lastRowInParent = parentPartitionBoundaries(parent + 1) // technically, first row in next IRM partition
+
+        val firstIndexToCopy = math.max(0, firstRowInBlock - firstRowInParent).toInt
+        val lastIndexToCopy = (math.min(lastRowInBlock, lastRowInParent) - firstRowInParent).toInt
+        
+        val indexedRowsInParent = irm.rows.iterator(parentPartitions(parent), context)
```

Comment:
You're computing the parent once for each block.  You should write out all the blocks at once.